src/js/trove/matrix.js

```diff
@@ -0,0 +1,1310 @@
+            for (var i = startpos; i < startpos + self.$w; i++) {
+                toret.$underlyingMat[startpos] = f(toret.$underlyingMat[startpos]);
+            }
+        }
+        //Applies a function to the given col
+        function internalColMap(self, col, f) {
+            const startpos = get1dElem(self, 0, col);
+            for (var i = 0; i < self.$h; i++) {
+                toret.$underlyingMat[get1dElem(self, i, col)] = f(toret.$underlyingMat[get1dElem(self, i, col)]);
+            }
+        }
+
+        // Adds self and other
+        var addMatrix = function (self, other) {
+            runtime.ffi.checkArity(2, arguments, "add-mat", false);
+            runtime.checkArgsInternal2("matrix", "add-mat", self, annMatrix, other, annMatrix);
+            if (!sameDims(self, other)) {
+                return runtime.ffi.throwMessageException("Matrices have dimensions " + printDims(self) + " and " + printDims(other) + " . They cannot be added");
+            } else {
+                new_arr = new Array(self.$l);
+                for (var i = 0; i < self.$l; i++) {
+                    new_arr[i] = runtime.plus(self.$underlyingMat[i], other.$underlyingMat[i]);
+                }
+
+                return createMatrixFromArray(self.$h, self.$w, new_arr);
+            }
+        };
+        //subtract matrix
+        var subMatrix = function (self, other) {
+            runtime.ffi.checkArity(2, arguments, "sub-mat", false);
+            runtime.checkArgsInternal2("matrix", "sub-mat", self, annMatrix, other, annMatrix);
+            if (!sameDims(self, other)) {
+                return runtime.ffi.throwMessageException("Matrices have dimensions " + printDims(self) + " and " + printDims(other) + " . They cannot be added");
+            } else {
+                new_arr = new Array(self.$l);
+                for (var i = 0; i < self.$l; i++) {
+                    new_arr[i] = runtime.minus(self.$underlyingMat[i], other.$underlyingMat[i]);
+                }
+
+                return createMatrixFromArray(self.$h, self.$w, new_arr);
+            }
+        }
+
+        //multiply matrix
+        var multMatrix = function (self, other) {
+            runtime.ffi.checkArity(2, arguments, "mult-mat", false);
+            runtime.checkArgsInternal2("Matrix", "mult-mat", self, annMatrix, other, annMatrix);
+            if (self.$w != other.$h) {
+                return runtime.ffi.throwMessageException("The width of the first matrix and the height of the second matrix need to be equal");
+
+            } else {
+                new_arr = new Array((self.$h * other.$w));
+                for (var i = 0; i < self.$h; i++) {
+                    for (var j = 0; j < other.$w; j++) {
+                        var elm = runtime.makeNumber(0);
+                        for (var k = 0; k < self.$w; k++) {
+                            elm = runtime.plus(elm, runtime.times(get1dElem(self, i, k), get1dElem(other, k, j)));
```

Comment:
I'm not sure whether you want to be using `runtime.plus` et al, or whether you should use the jsNumbers library directly.  Any error messages you get will have bizarre call-stacks if you do it this way.

Reply:
This triple-loop is a potential stall-point for a Pyret program.  If I have two big matrices, and I multiply them, the multiplication could take long enough that it stalls the browser.  You may need to chunk this up via `safeCall`s into smaller pieces.